rfd/0156-notifications.md

```diff
@@ -0,0 +1,378 @@
+- `default-warning`: The default subkind for a warning notification.
+- `user-created-informational`: The subkind for informational notifications that were manually created by a user using `tctl`.
+- `user-created-warning`: The subkind for warning notifications that were manually created by a user using `tctl`.
+- `access-request-pending`: The subkind for a notification for an access request pending review. This is the notification received by a user when an access request needs their review.
+- `access-request-approved`: The subkind for a notification for a user's access request being approved.
+- `access-request-denied`: The subkind for a notification for a user's access request being denied.
+
+For example, if a notification's subkind is `access-request-pending`, the UI will make the notification redirect to the page for reviewing the access request.
+
+#### Usage
+
+The notifications service will work primarily by adding, deleting, and fetching notifications from the backend.
+
+These are some examples of methods the service will expose:
+
+- `NotifyUser` - Sends a notification to a specific user based on the specified username.
+- `NotifyUserByPermissions` - Sends a notification to all users with the specified permission(s).
+- `NotifyUsersByRole` - Sends a notification to all users with the specific role(s).
+- `NotifyAllUsers` - Sends a notification to all users.
+- `GetNotificationsForUser` - Returns all the notifications for a user after filtering.
+- `UserMatchesGlobalNotification` - Checks if a user matches a global notification's matcher.
+- `DeleteGlobalNotification` - Deletes a global notification.
+- `SetUserLastSeenNotification` - Sets the timestamp of the last notification a user has seen.
+- `MarkNotificationAsClicked` - Marks a notification as having been clicked on.
+- `MarkNotificationAsDismissed` - Marks a notification as dismissed.
+
+##### Example Usage:
+
+```go
+
+// Notifies all users with the permission to `create` and `update` integrations.
+
+roleConditions := []types.RoleConditions{
+   {
+      Rules: []types.Rule{
+         {
+            Resources: []string{types.KindIntegration},
+            Verbs:     []string{types.VerbCreate, types.VerbUpdate},
+         },
+      },
+   },
+}
+
+notification := types.Notification{
+   Type: notifications.IntegrationMisconfiguredType,
+   Metadata: types.Metadata{
+      Name: "Integration Misconfigured",
+      Description: fmt.Sprintf("The integration [%s] was misconfigured.", integration.name),
+      Labels:      map[string]string{"integration-id": integration.id},
+   },
+}
+
+notifyUsersByPermissions(roleConditions, notification)
+```
+
+### Flow
+
+#### Web UI
+
+Upon logging in and on every page load, a request will be made to fetch all the user's notifications. This fetch will be performed asynchronously so that the login or page loading process is not slowed down by having to wait for the notifications response.
```

Comment:
In case of multiple simultaneous requests processing we will abort any previous ones and only process the most recent one, since this type of handling is standard I didn't think it was worth mentioning in the RFD.

Reply:
Good idea, I'll add this to the RFD.